Task graph diagnostic (#33095)
gradle/gradle: platforms/documentation/docs/src/docs/release/notes.md

For more details and potential breakages, see the dedicated [upgrading guide section](userguide/upgrading_version_8.html).

### Task graph diagnostic

> We introduce a new task graph diagnostic feature to help you understand the task graph of your build.

Reviewer: ```suggestion
Gradle 9.0.0 introduces a new task graph diagnostic feature to help you visualize the task execution plan without running any tasks.
```

Author: Note: tasks from build logic still can run, so "without running any tasks" is misleading

> You can use `--task-graph` command line option to print the task graph of your build without executing the tasks, for example:

Reviewer: ```suggestion
You can enable it using the `--task-graph` command-line option. For example:
```

```
$ ./gradlew root r2 --task-graph

> 

Reviewer: Please remove the `//` below, the goal is to seperate the example into two code snippets (instead of one):

```suggestion
//```
This prints a visual representation of the task graph for the specified tasks:
//```
```

Author: Lifehack for the future: you can use more than three ticks to have nested code blocks. Like this:
`````
````suggestion
some words
```kotlin
some code
```
````
`````

> Tasks graph for: root r2

Reviewer: This should be a seperate code snippet:

```suggestion
//```
Tasks graph for: root r2
```

+--- :root (org.gradle.api.DefaultTask)
|    \\--- :middle (org.gradle.api.DefaultTask)
|         +--- :leaf1 (org.gradle.api.DefaultTask)
|         \\--- :leaf2 (org.gradle.api.DefaultTask, disabled)
\\--- :root2 (org.gradle.api.DefaultTask)
    +--- :leaf1 (org.gradle.api.DefaultTask) (*)
    |--- other build task :included:fromIncluded (org.gradle.api.DefaultTask)
    \\--- :leaf4 (org.gradle.api.DefaultTask, finalizer)
         \\--- :leaf3 (org.gradle.api.DefaultTask)
         
(*) - details omitted (listed previously)
```

> This feature can be useful to get a quick overview of the task graph and to understand the dependencies between tasks.

Reviewer: Feel free to change this but it would be nice as a list:
```suggestion
This feature is especially useful for:
- Gaining a quick overview of task dependencies
- Debugging why a task is scheduled
- Exploring subgraphs by refining your task selection
```

Author: I think these items are from different categories, so it feels a bit weird to me to put them on one list.

Reviewer: Let's do this instead:
```suggestion
This feature provides a quick overview of the task graph, helping users understand the dependencies between tasks without running them.
```

> You can iterate by diving into a subgraph by adjusting an invocation.

Reviewer: ```suggestion
```

Reviewer: If we want to keep this, it needs further explanation.

> Note that this feature is still in incubation and may change in future releases.

Reviewer: ```suggestion
This feature is incubating and may change in future releases.
```

> Also, there is a known [issue](https://github.com/gradle/gradle/issues/2517) about the behavior of the `--dry-run` option with included builds: tasks from included builds will be executed.

Reviewer: ```suggestion
Additionally, it shares a known [issue](https://github.com/gradle/gradle/issues/2517) with `--dry-run`:
Tasks from included builds may still be executed. 
```

> This feature suffers from the same issue.

Reviewer: ```suggestion
```

-->

### Task graph diagnostic

> Gradle 9.0.0 introduces a new task graph diagnostic feature to help you visualize the task execution plan without running tasks.

Reviewer: ```suggestion
A new task dependency graph is available to visualize the dependencies between tasks without executing them. 
```

> You can enable it using the `--task-graph` command-line option. For example:

Reviewer: ```suggestion
You can enable it using the `--task-graph` option on the command line. For example:
```

```

> ./gradlew root r2 --task-graph

Author: 🤔 I intentionally chose abstract names to avoid any attachment to real tasks to demonstrate the shape / appearance of the graph. It also shows finalizers and included builds; it would be hard to find a real-world example with all of that.

> $ ./gradlew root r2 --task-graph

Reviewer: ```suggestion
./gradlew root r2 --task-graph
```
